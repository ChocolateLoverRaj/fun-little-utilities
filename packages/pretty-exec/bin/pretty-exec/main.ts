import { spawnSync } from 'child_process'
import process from 'process'
import { main } from '../../index'

const status = main({
  spawn: spawnSync,
  print: console.info,
  error: console.error,
  process,
})

if (status === null) {
  console.error('[ERROR] Exit status is null, something goes wrong')
  throw process.exit(-1)
}

throw process.exit(status)
