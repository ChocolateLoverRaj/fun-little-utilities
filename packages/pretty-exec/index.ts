export * from './lib/chalk'
export * from './lib/parse-command-arguments'
export * from './lib/pretty-exec'
export * from './lib/cli'
export { default } from './lib/pretty-exec'
