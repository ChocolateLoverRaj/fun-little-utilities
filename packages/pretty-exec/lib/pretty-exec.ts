import chalk from 'chalk'
import parseCommandArguments, { ParsingResult } from './parse-command-arguments'

export interface SpawnOptions {
  stdio: 'inherit'
}

export interface SpawnFunc<Return> {
  (command: string, args: string[], options: SpawnOptions): Return
}

export interface Printer {
  (...args: string[]): void
}

export interface StylingFunc {
  (text: string): string
}

export interface ColorSchemes {
  readonly prefix?: StylingFunc
  readonly command?: StylingFunc
  readonly argument?: StylingFunc
  readonly flag?: StylingFunc
  readonly optionKey?: StylingFunc
  readonly optionValue?: StylingFunc
}

export interface CreateOptions<SpawnReturn> {
  readonly spawn: SpawnFunc<SpawnReturn>
  readonly print: Printer
  readonly prefix?: readonly string[]
  readonly colorSchemes?: ColorSchemes
}

export interface PrettyExec<Return> {
  (command: string, ...args: string[]): Return
}

export const DEFAULT_PREFIX = ['$'] as const
export const DEFAULT_COLOR_SCHEMES: ColorSchemes = {}
export const DEFAULT_PREFIX_STYLING_FUNC: StylingFunc = chalk.dim
export const DEFAULT_ARGUMENT_STYLING_FUNC: StylingFunc = chalk
export const DEFAULT_COMMAND_STYLING_FUNC: StylingFunc = chalk.green
export const DEFAULT_FLAG_STYLING_FUNC: StylingFunc = chalk.red
export const DEFAULT_OPTION_KEY_STYLING_FUNC: StylingFunc = chalk.red
export const DEFAULT_OPTION_VALUE_STYLING_FUNC: StylingFunc = chalk
export const SPAWN_OPTIONS: SpawnOptions = { stdio: 'inherit' }

export function createPrettyExec<Return> (options: CreateOptions<Return>): PrettyExec<Return> {
  const {
    spawn,
    print,
    prefix: prefixText = DEFAULT_PREFIX,
    colorSchemes = DEFAULT_COLOR_SCHEMES
  } = options

  const {
    prefix = DEFAULT_PREFIX_STYLING_FUNC,
    command = DEFAULT_COMMAND_STYLING_FUNC,
    argument = DEFAULT_ARGUMENT_STYLING_FUNC,
    flag = DEFAULT_FLAG_STYLING_FUNC,
    optionKey = DEFAULT_OPTION_KEY_STYLING_FUNC,
    optionValue = DEFAULT_OPTION_VALUE_STYLING_FUNC
  } = colorSchemes

  return (cmd, ...args) => {
    print(
      ...prefixText.map(text => prefix(text)),
      command(cmd),
      ...parseCommandArguments(args).map(item => {
        switch (item.type) {
          case ParsingResult.Type.Argument:
            return argument(item.text)
          case ParsingResult.Type.Flag:
            return flag(item.text)
          case ParsingResult.Type.Option:
            return optionKey(item.key) + optionValue('=' + item.value)
        }
      })
    )

    return spawn(cmd, args, SPAWN_OPTIONS)
  }
}

export default createPrettyExec
