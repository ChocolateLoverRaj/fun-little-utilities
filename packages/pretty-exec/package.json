{
  "name": "pretty-exec",
  "version": "0.0.0",
  "description": "Print a command and then execute it",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/ksxnodeapps/fun-little-utilities#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "bin": {
    "pretty-exec": "bin/index",
    "print-and-execute": "bin/index"
  },
  "dependencies": {
    "chalk": "^2.4.2",
    "split-string-once": "0.0.0",
    "tslib": "^1.10.0",
    "@types/node": "^12.7.5"
  }
}
