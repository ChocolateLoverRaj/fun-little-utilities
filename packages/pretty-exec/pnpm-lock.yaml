dependencies:
  '@types/node': 12.7.5
  chalk: 2.4.2
  split-string-once: 0.0.0
  tslib: 1.10.0
lockfileVersion: 5.1
packages:
  /@types/node/10.14.18:
    dev: false
    resolution:
      integrity: sha512-ryO3Q3++yZC/+b8j8BdKd/dn9JlzlHBPdm80656xwYUdmPkpTGTjkAdt6BByiNupGPE8w0FhBgvYy/fX9hRNGQ==
  /@types/node/12.7.5:
    dev: false
    resolution:
      integrity: sha512-9fq4jZVhPNW8r+UYKnxF1e2HkDWOWKM5bC2/7c9wPV835I0aOrVbS/Hw/pWPk2uKrNXQqg9Z959Kz+IYDd5p3w==
  /ansi-styles/3.2.1:
    dependencies:
      color-convert: 1.9.3
    dev: false
    engines:
      node: '>=4'
    resolution:
      integrity: sha512-VT0ZI6kZRdTh8YyJw3SMbYm/u+NqfsAxEpWO0Pf9sq8/e94WxxOpPKx9FR1FlyCtOVDNOQ+8ntlqFxiRc+r5qA==
  /chalk/2.4.2:
    dependencies:
      ansi-styles: 3.2.1
      escape-string-regexp: 1.0.5
      supports-color: 5.5.0
    dev: false
    engines:
      node: '>=4'
    resolution:
      integrity: sha512-Mti+f9lpJNcwF4tWV8/OrTTtF1gZi+f8FqlyAdouralcFWFQWF2+NgCHShjkCb+IFBLq9buZwE1xckQU4peSuQ==
  /color-convert/1.9.3:
    dependencies:
      color-name: 1.1.3
    dev: false
    resolution:
      integrity: sha512-QfAUtd+vFdAtFQcC8CCyYt1fYWxSqAiK2cSD6zDB8N3cpsEBAvRxp9zOGg6G/SHHJYAT88/az/IuDGALsNVbGg==
  /color-name/1.1.3:
    dev: false
    resolution:
      integrity: sha1-p9BVi9icQveV3UIyj3QIMcpTvCU=
  /escape-string-regexp/1.0.5:
    dev: false
    engines:
      node: '>=0.8.0'
    resolution:
      integrity: sha1-G2HAViGQqN/2rjuyzwIAyhMLhtQ=
  /has-flag/3.0.0:
    dev: false
    engines:
      node: '>=4'
    resolution:
      integrity: sha1-tdRU3CGZriJWmfNGfloH87lVuv0=
  /split-string-once/0.0.0:
    dependencies:
      '@types/node': 10.14.18
      tslib: 1.10.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-FrmRtP56cdaHEPAGGjV6eebDJ7QUZ+sobv2SV69Fec5G6cJ2Jwm1d8o+IE1HyweCPYz6Mf7WVl/dpSkuN6tUbQ==
  /supports-color/5.5.0:
    dependencies:
      has-flag: 3.0.0
    dev: false
    engines:
      node: '>=4'
    resolution:
      integrity: sha512-QjVjwdXIt408MIiAqCX4oUKsgU2EqAGzs2Ppkm4aQYbjm+ZEWEcW4SfFNTr4uMNZma0ey4f5lgLrkB0aX0QMow==
  /tslib/1.10.0:
    dev: false
    resolution:
      integrity: sha512-qOebF53frne81cf0S9B41ByenJ3/IuH8yJKngAX35CmiZySA0khhkovshKK+jGCaMnVomla7gVlIcc3EvKPbTQ==
specifiers:
  '@types/node': ^12.7.5
  chalk: ^2.4.2
  split-string-once: 0.0.0
  tslib: ^1.10.0
