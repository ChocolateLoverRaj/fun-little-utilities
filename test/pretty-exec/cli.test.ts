import { main } from 'pretty-exec'

function setup (argv: readonly string[]) {
  const status = 123
  const history = Array<any>()
  const spawn = jest.fn(() => {
    history.push('spawn')
    return { status } as const
  })
  const print = jest.fn(() => {
    history.push('print')
  })
  const error = jest.fn(() => {
    history.push('error')
  })
  const result = main({
    process: { argv } as const,
    spawn,
    print,
    error
  })
  return { status, history, spawn, print, error, argv, result }
}

describe('when argv is sufficient', () => {
  const argv = ['node', 'script', 'command', 'abc', 'def'] as const

  it('calls spawn once', () => {
    const { spawn } = setup(argv)
    expect(spawn).toBeCalledTimes(1)
  })

  it('calls spawn with expected arguments', () => {
    const { spawn } = setup(argv)
    const [command, ...rest] = argv.slice(2)
    expect(spawn).toBeCalledWith(command, rest, { stdio: 'inherit' })
  })

  it('calls print once', () => {
    const { print } = setup(argv)
    expect(print).toBeCalledTimes(1)
  })

  it('calls print with expected arguments', () => {
    const { print } = setup(argv)
    expect(print.mock.calls).toMatchSnapshot()
  })

  it('does not call error', () => {
    const { error } = setup(argv)
    expect(error).not.toBeCalled()
  })

  it('returns status code', () => {
    const { status, result } = setup(argv)
    expect(result).toBe(status)
  })

  it('calls print before spawn', () => {
    const { history } = setup(argv)
    expect(history).toEqual(['print', 'spawn'])
  })
})

describe('when argv is insufficient', () => {
  const argv = ['node', 'script']

  it('does not call spawn', () => {
    const { spawn } = setup(argv)
    expect(spawn).not.toBeCalled()
  })

  it('does not call print', () => {
    const { spawn } = setup(argv)
    expect(spawn).not.toBeCalled()
  })

  it('calls error once', () => {
    const { error } = setup(argv)
    expect(error).toBeCalledTimes(1)
  })

  it('calls error with a message', () => {
    const { error } = setup(argv)
    expect(error).toBeCalledWith('[ERROR] Command is missing')
  })

  it('returns -1', () => {
    const { result } = setup(argv)
    expect(result).toBe(-1)
  })
})

describe('when argv is just enough', () => {
  const argv = ['node', 'script', 'command'] as const

  it('calls spawn once', () => {
    const { spawn } = setup(argv)
    expect(spawn).toBeCalledTimes(1)
  })

  it('calls spawn with expected arguments', () => {
    const { spawn } = setup(argv)
    const [command, ...rest] = argv.slice(2)
    expect(spawn).toBeCalledWith(command, rest, { stdio: 'inherit' })
  })

  it('calls print once', () => {
    const { print } = setup(argv)
    expect(print).toBeCalledTimes(1)
  })

  it('calls print with expected arguments', () => {
    const { print } = setup(argv)
    expect(print.mock.calls).toMatchSnapshot()
  })

  it('does not call error', () => {
    const { error } = setup(argv)
    expect(error).not.toBeCalled()
  })

  it('returns status code', () => {
    const { status, result } = setup(argv)
    expect(result).toBe(status)
  })

  it('calls print before spawn', () => {
    const { history } = setup(argv)
    expect(history).toEqual(['print', 'spawn'])
  })
})
