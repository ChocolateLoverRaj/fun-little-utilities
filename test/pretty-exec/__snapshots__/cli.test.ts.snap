// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when argv is sufficient calls print with expected arguments 1`] = `
Array [
  Array [
    "[2m$[22m",
    "[32mcommand[39m",
    "abc",
    "def",
  ],
]
`;
